lineinfile file locking support

acalm opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Added support for file locking through common.file.FileLock (fixes #30413). Also fixed issue where FileLock.set_lock in module_utils/common/file.py didn't properly handle the default value None
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
lineinfile
module_utils/common/file.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.8.0.dev0 (fix-30413 2455b96cc1) last updated 2018/10/19 00:04:07 (GMT +200)
  config file = None
  configured module search path = ['/home/acalm/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/acalm/venv/ansible/lib/python3.6/site-packages/ansible
  executable location = /home/acalm/venv/ansible/bin/ansible
  python version = 3.6.5 (default, Apr 20 2018, 08:08:06) [GCC 6.4.0]
```

##### ADDITIONAL INFORMATION
N/A

Commit:
implement locking for lineinfile

Commit:
update fragment

Commit:
properly handle None as lock_timeout

Commit:
fix pep8/ansible validation errors

samdoran commented on changelogs/fragments/30413-lineinfile-concurrence_issue.yaml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+- "Fixed lineinfile concurrence issue with multiple delegation on same delegated host and same file"
+- "common.file.FileLocking now properly handles lock_timeout = None"
```
Change this entry to state that "`common.file.FileLock.set_lock()` now handles negative values the same as zero". The `set_lock()` method seems to already handle `None` properly, but not negative values.

samdoran replied:
Hmm, actually, I'm not sure this change is correct.

acalm replied:
Will do!
Actually the lock_timeout <= 0 bailed out when None, however now it handles both negative and None, whereas before it behaved a bit unexpected. I'll set up some integration tests as soon as possible!

samdoran reviewed (changes requested):
Can you please add integration tests for this new feature?

samdoran commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -89,7 +89,7 @@ def set_lock(self, path, tmpdir, lock_timeout=None):
         self.lockfd = open(lock_path, 'w')
 
-        if lock_timeout <= 0:
+        if lock_timeout == 0:
```
Since the `lineinfile` defaults to `-1` for the `lock_timeout`, this conditional is skipped but the next matches and an exception is raised. Since `0` is less than `-1` but `-1` is "truthy", the while loop is skipped, the lockfile is immediately closed, and an exception as raised.

I believe setting the default for `lock_timeout` to 0 or possibly a small positive integer in the `argument_spec` would be the best solution. Then, this change would not be necessary.

samdoran replied:
Another option would be to change the default of `lock_timeout` to `0` in the method signature.

acalm replied:
Ah, you're right!
Problem here is that in python3 using `lock_timeout <= 0` throws a TypeError exception  ('<=' not supported between instances of 'NoneType' and 'int') if the default value is left as is. It however works fine on python2, which of course was used while testing FileLock. Anyway, at the time it seemed more logical to have None as default for lock_timeout as it indicates that timeout shouldn't be used at all, I.E wait forever, this however mixes types (None and int) which is a bit inelegant and makes the ansible-validation angry. I was thinking that 0 should probably indicate that the timeout should be 0 and fail immediately if a lock is blocking.

Next if could probably be changed to `if lock_timeout > 0:` to handle both int and None properly?

Commit:
negative lock_timeout should be treated as disabled

dagwieers commented:
I would prefer we fix this for all our modules, rather than do it for a single module.
There's a proposal for this at: https://github.com/ansible/proposals/issues/113

The nice thing about using flock() is that the lock automatically disappears with the process.

dagwieers commented:
@acalm So what you are doing is fine from the module-level point of view, but what I think is more worthwhile is that we have wrapper-functions for various file-level operations that make it easier for any module to support flock. So rather than doing this one-time for the **lineinfile** module, we should be looking at what is needed for doing this for every module in a more concerted effort.

Commit:
add integration test for lineinfile using lock

acalm commented:
ready_for_review

dagwieers reviewed (changes requested):
As discussed earlier, I don't think we want to add this on a per-module basis. It means we'll be adding lock-related parameters to individual modules, whereas IMO this should be handled from module_utils by having a framework and global/task-specific vars, not parameters.

There are plenty of modules that would benefit from this (anything that updates remote files).

Commit:
really accept None as lock_timeout

Commit:
pep8 fix

dagwieers commented:
@sivel @bcoca I would like to know how we want to proceed with this. I'd rather not have parameters on a per-module basis, but a general flag that can be enabled/disabled either globally or on the task-level. Also see https://github.com/ansible/proposals/issues/113

bcoca commented:
@dagwieers I'm unsure, a 'task level' parameter might make sense for other modules that don't deal with files but might want to handle concurrency anyways, but I'm hard pressed to come up with good examples.  FYI, the current implementation relies on 'flag files' and not on the original target file.

Adding @abadger as he oversaw the current lockfile common code.

Commit:
use module.tmpdir instead of system default tmpdir

abadger commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -45,7 +45,7 @@ def is_executable(path):
 
 
-class FileLock:
+class FileLock(object):
```
This isn't needed because of the ```__metaclass__ =  type``` boilerplate

abadger commented:
It's been a while so I don't remember all of the discussion that may have gone into this before but I'm unsure why we want to expose locking to the end user here?  Shouldn't it just be something that the lineinfile does unconditionally?  Maybe (I'm not sure about this) timeout could be exposed as some generic timeout but even that I'm not sure about.

Commit:
remove unnecessary object

Commit:
forgot ()

Commit:
remove lock arguments, use FileLock by default

acalm commented:
There's probably no downsides enabling locking by default here commit 043aeef removes the module attributes and wraps the present and absent actions within the FileLock lock_file context.

Commit:
handle utf-8 file names

dagwieers commented:
@abadger Exactly, I would like over time that all modules that can take advantage of this have locking (always) enabled so that each can safely modify files without concurrent sessions step on each other's toes. The only reason for having a (global/local) way to influence this was that there were concerns about performance (and a need to enable/disable this).

Performance impact should be negligible, unless in those cases where it matters (concurrent access), so I would prefer we simply add this to more modules, and make sure we have integration tests covering this (so we probably need an async way of locking a file for a period of time for integration tests).

Maybe a global timeout configuration value could be useful, but it means we have to expose this to the module. Who can help make a decision about this so we can go forward and add this to some more modules ? (I have a need for this since some time myself...)

dagwieers commented:
Also, the proposal ansible/proposals#113 was going to use the file to be modified as the lock-file, so that it doesn't matter what user wants to acquire the lock, it would still work. Using a lock-file in a temporary location would only work for access from Ansible with that specific user.

A more generic solution would work for other tools using the same convention.

acalm commented:
@dagwieers we briefly discussed the multi/single -user aspect and the use of the hybrid lock file model, (flock a lockfile) while creating the common.file.FileLock. I vaguely recall that locking the actual file could/would/maybe lead to some weird race conditions or other weird behaviour, sadly both my memory and irc logs from then is gone :(. Putting the log file in a user-specific temp dir was made to prevent a ddos scenario, I can't remember the exact scenario but I can dig it up from logs. I would also like to improve filelock to better fit ansible/proposals#113

Commit:
revert a376afe00dac3e0f66f9a279b29c2980ed38775d

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/91889/66/tests

mattclay commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -470,7 +471,7 @@ def main():
             backup=dict(type='bool', default=False),
             firstmatch=dict(default=False, type='bool'),
-            validate=dict(type='str'),
+            validate=dict(type='str')
```
Keep the trailing comma. Having it there makes for a cleaner diff if another line is added below.

Commit:
lock_file to_bytes and trailing comma

tuxknight commented:
any updates on this PR ?

dagwieers commented:
This was discussed during the core meeting at: https://meetbot.fedoraproject.org/ansible-meeting/2018-11-06/ansible_core_public_irc_meeting.2018-11-06-19.04.log.html starting at `
20:20:58 <bcoca> #topic https://github.com/ansible/ansible/pull/47322`.

To summarize:
- This PR fixes some bugs in `set_lock()` and makes the locking transparent to the user
  - Having these fixes in a separate PR would likely speed up getting these fixes in Ansible already

- There is a wish to have this preferably work automatically (so without the need of task/module options)
  - We need to know the actual cost of locking implicitly
  - If the implementation makes sense for the problem cases in which we want automatic locking

- Do we want locking on the actual file, or not?
  - If we do the locking on a temporary file it will be very Ansible-specific and would only work for access as the same user, while if we lock the actual file other tools using the same convention would work with Ansible
  - There was a mention of a possible race condition when locking the actual file, but I'd like to understand why that is on the real file, and not on a temporary file

dagwieers reviewed (approved):
I personally wouldn't mind getting this merged in full, with the knowledge that we may want to improve on this with a broader scope in mind.

dagwieers commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -194,6 +194,7 @@
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils.six import b
 from ansible.module_utils._text import to_bytes, to_native
+from ansible.module_utils.common.file import FileLock
```
We tend to sort the imports by name.

acalm replied:
fixed in commit b7d653a

acalm commented:
@dagwieers thanks for the summarization.I've been thinking about how to change the current implementation to better fit [ansible/proposals#113](https://github.com/ansible/proposals/issues/113) but I haven't been around irc much to discuss this, I seem to be in the wrong timezone :).

>     * There is a wish to have this preferably work automatically (so without the need of task/module options)
I think automatic locking will be rather tricky unless we somehow hijack pythons open(), because the lock should be placed before there's any operations against the file (as mentioned [here](https://github.com/ansible/proposals/issues/113#issuecomment-387898356)).

>     * Do we want locking on the actual file, or not?
>       
>       * If we do the locking on a temporary file it will be very Ansible-specific and would only work for access as the same user, while if we lock the actual file other tools using the same convention would work with Ansible
>       * There was a mention of a possible race condition when locking the actual file, but I'd like to understand why that is on the real file, and not on a temporary file

I can't remember the specific race condition or issue that was discussed, however while working with another project that had a lot of threads doing file operations there where sometimes issues where one process changed permissions on one file while another process, owning the lock, tried to write and would fail, this might've been the reason. However, I believe there shouldn't be a problem to change the current behaviour to just use the actual file, the chmod part needs to be removed from file_lock() though.

And yea, in retrospect I probably should've made two different pull requests.

dagwieers commented:
> I think automatic locking will be rather tricky unless we somehow hijack pythons open(), because the lock should be placed before there's any operations against the file (as mentioned here).

I think the correcter term would have been unconditional locking (I don't think we want to hijack lower-level operations, as there's no context that e.g. a read is done with the intent of rewriting the file).

> However, I believe there shouldn't be a problem to change the current behaviour to just use the actual file, the chmod part needs to be removed from file_lock() though.

By using `module.tmpdir` currently 2 different concurrent tasks would happily access and write concurrently to the same file, as they use a different lock. So AFAICT the existing implementation will not work.

acalm commented:
> By using module.tmpdir currently 2 different concurrent tasks would happily access and write concurrently to the same file, as they use a different lock. So AFAICT the existing implementation will not work.
It will work if the 2 concurrent tasks are executed by the same user, there's a comment about this quirk in file.py

dagwieers commented:
> It will work if the 2 concurrent tasks are executed by the same user, there's a comment about this quirk in file.py

I am afraid that's not true, every task has a unique `tmpdir` value, even for the same user. To demonstrate I made a small module that returns `module.tmpdir`. And used the following playbook:
```yaml
- hosts: localhost
  tasks:
  - shell: whoami
  - test:
    loop:
    - one
    - two
  - test:
  - test:
```
And this is the output:
```json
PLAY [localhost] ***********************************************************************************************

TASK [test] ****************************************************************************************************
ok: [localhost] => (item=one) => {"changed": false, "item": "one", "tmpdir": "/home/dag/.ansible/tmp/ansible-tmp-1543612467.92-118739727765684/"}
ok: [localhost] => (item=two) => {"changed": false, "item": "two", "tmpdir": "/home/dag/.ansible/tmp/ansible-tmp-1543612468.21-4113821783417/"}

TASK [test] ****************************************************************************************************
ok: [localhost] => {"changed": false, "tmpdir": "/home/dag/.ansible/tmp/ansible-tmp-1543612468.39-84404110220373/"}

TASK [test] ****************************************************************************************************
ok: [localhost] => {"changed": false, "tmpdir": "/home/dag/.ansible/tmp/ansible-tmp-1543612468.58-148180724190992/"}

PLAY RECAP *****************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0
```

Commit:
fix import order and revert changes in 220280b

Commit:
put lock on the actual file, avoid tmpdir

acalm commented:
@dagwieers I've updated FileLock to lock the actual file instead, this should better fit the proposal and work with multiple users/processes. I'm not sure where the lock timeout should live though, some setting in ansible.cfg or somewhere else?

Commit:
use ab, remove unused lock_path variable

dagwieers commented:
@acalm So that's one point for discussion. I would take a sensible default (what is sensible in our context ?) but allow the possibility to increase that value per task (and maybe 0 or -1 means that it is disabled ?).

Would a value of 10 or 15 seconds be a safe value, or instead take something like 30 seconds ? What operations in this case would take longer than this ? Would it make sense to allow modules to influence the timeout value because some modules are heavier than others ? This could provide a better user experience (when there are concurrency-related issues).

dagwieers commented:
Some references on file locking on Unix:
- http://0pointer.de/blog/projects/locking.html
- https://unix.stackexchange.com/questions/174955/generic-shared-resource-locking-techniques-is-flock-the-solution
- http://chris.improbable.org/2010/12/16/everything-you-never-wanted-to-know-about-file-locking/

So it seems the best portable option is using python's `fnctl.lockf()` implementation (which apparently is using `fnctl()` and not `lockf()` internally)

dagwieers commented:
The tests fail because some files don't exist during testing.
```
  File "/tmp/ansible_lineinfile_payload_UXnveB/ansible_lineinfile_payload.zip/ansible/module_utils/common/file.py", line 64, in lock_file
  File "/tmp/ansible_lineinfile_payload_UXnveB/ansible_lineinfile_payload.zip/ansible/module_utils/common/file.py", line 88, in set_lock
IOError: [Errno 2] No such file or directory: '/tmp/ansible-local-\xe6\xb1\x89\xe8\xaf\xad/\xe6\xb1\x89\xe8\xaf\xad.txt'
```

Commit:
use lockf instead of flock

Commit:
move locking into present and absent to avoid 'No such file or directory' errors

Commit:
fcntl.lockf() have unusual side effects

acalm commented:
Avoid lockf quirk, close an unrelated file descriptor to the same file as the lock, the lock is released.

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -251,9 +254,12 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
-            b_lines = f.readlines()
+
+    if not module.check_mode:
+        flock.set_lock(dest)
```
It's better to do this as a context manager so it gets released once we leave the context manager's scope.

Since we read the file outside of check mode, we probably want to lock outside of check mode as well.

dagwieers replied:
My feeling is that we only need to lock a file if we read with the intention of rewriting that file. So that others don't read an older version for writing after someone else has written. We don't mind if other instances read older versions, because that's just timing.

acalm replied:
I wrapped present() and absent() in a context manager before, but inside main() which fails, because sometimes the destination path haven't been created yet. I thought about using a context manager from within the functions, it would work but nearly everything inside the function(s) will be wrapped inside of it. Also it would put a lock on the file even when running with --check.
The lock will be released automatically when the process dies, unlock isn't necessary. I agree it's better to do it with a context manager, here however it's inconvenient.

abadger commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,57 @@ class FileLock:
         :returns: True
         '''
-        lock_path = os.path.join(tmpdir, 'ansible-{0}.lock'.format(os.path.basename(path)))
+        lock_path_b = to_bytes(path, errors='surrogate_or_strict')
```
Note: Our convention is to use a prefix (b_locak_path) rather than a suffix.

Commit:
use context manager

Commit:
fix pep8 concerns

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -250,10 +248,27 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
+    b_dest = to_bytes(dest, errors='surrogate_or_strict')
+
+    with open(b_dest, 'rb') as f:
+        b_lines = f.readlines()
```
Looking good.  I'd do a little more reorganization of the content, though.

* Move these lines for reading from the file into ```present()```
* Move the ```write_changes()``` call into ```present()```.
* Move the module.exit_json() into ```present()``` (or higher, into ```main()```, but I can see that it  was in present() before so it's a preexisting problem).
* Change what we're passing in to ```_present_data_manipulator``` so that we do not pass in things that have to do with the file (module, dest, and backup), but it does include b_lines.
* Assembling the diff results and checking file attributes probably gets moved out to the higher level as well, although I'm not entirely happy with that.... those seem to float in the middle of the overall organization that we're shooting for.
* (For a separate PR) we can also change ```_present_data_manipulator()``` to return a copy of b_lines instead of modifying the data in place.  In place uses less memory but a function without side effects is generally better code. (As an example, think about the additional flexibility of python's ```sorted()``` function over ```$iterable.sort()```

The object is to make this helper function be only about manipulating the data and returning the new data.  Everything about the file itself or the fact that we're part of an Ansible module rather than some other script is removed from the helper.  This is a better organization for people to read in the future as all of the inputs and all of the outputs of ```_present_data_manipulator``` then become obvious to someone who is reading the code.

One way to judge this is to think of how you'd unit test this new function.  If you succeed in moving everything about file handling and module outside of the new function, then the new function is very straightforward to unittest.  You just pass in different values of b_lines (the data to be manipulated) and the options which change behaviour, regexp, line, insertbefore, insertafter, backrefs, and firstmatch, and then you see whether the lines that are output are what you expected or not.

By contrast, if the file manipulation and module is part of the new code, then you end up having to mock file objects and AnsibleModule in order to unit test the code.

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -250,10 +248,27 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
+        _present_data_manipulator(module, dest, regexp, line, insertafter, insertbefore,
+                                  backup, backrefs, firstmatch)
+
+    with flock.lock_file(dest):
```
missing an ```else:``` here.

acalm replied:
The else shouldn't™ be necessary (I think), dest will be created if it does not exist (if create=True) if the directory creation fails an exception will be thrown.

abadger replied:
Yeah, the problem case is no longer here.

Commit:
move non-data altering out of _present_data_manipulator

Commit:
move nested function and non-manipulative actions out of _absent_data_manipulator

abadger commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,57 @@ class FileLock:
             r_exception = BlockingIOError
 
-        self.lockfd = open(lock_path, 'w')
+        self.lockfd = open(b_lock_path, 'ab')
```
We'll need to change this to raise an IOError if the file does not exist.  With the previous code, this was okay because we were opening and locking a temporary file but now that we're operating on the real file we need to ensure that we don't end up creating a nonexistent file in check mode.

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -248,15 +252,37 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
             try:
                 os.makedirs(b_destpath)
```
Once the locking code is updated to raise IOError if the file does not exist, we'll have to create the file here.  I believe we can do that like:
``` python
open(b_dest, 'ab').close()
```

I believe using append mode, if the file doesn't exist, it will be created.  If the file does exist before the filehandle is closed, it won't be modified.

samdoran replied:
Verified this is the case: an existing file is not modified but an empty file is created if it does not exist using the above code.

abadger commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -248,15 +252,37 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
-    else:
+    with flock.lock_file(dest):
         with open(b_dest, 'rb') as f:
             b_lines = f.readlines()
```
Once we have the flock.lock_file() code raising an IOError if the file does not exist, we can rewrite this section like this:
``` python
try:
    with flock.lock_file(dest):
        with open(b_dest, 'rb') as f:
            b_lines = f.readlines()
        msg, changed, b_modified_lines = _present_data_manipulator(b_lines, regexp, line, insertafter,
                                                                   insertbefore, backrefs, firstmatch)
        if changed and not module.check_mode:
            if backup and os.path.exists(b_dest):
                backupdest = module.backup_local(dest)
            write_changes(module, b_modified_lines, dest)
except IOError as e:
    if not module.check_mode and 'file doesn't exist when locking' in to_native(e):
        raise
    # Check mode won't create the file so we know it starts off empty
    b_lines = []
    msg, changed, b_modified_lines = _present_data_manipulator(b_lines, regexp, line, insertafter,
                                                                   insertbefore, backrefs, firstmatch)

diff['before'] = to_native(b('').join(b_lines))
[...]
```

One thing I don't know is whether check_file_attrs should be inside of the lock. If it does, I think it can still be added in up there but there would need to be further rearrangement or another test of whether we're in check_mode or not.

acalm replied:
I wrapped it in if/else, to prevent locking while running in check_mode, I think it's cleaner even though there's some duplicated code. I left check_file_attrs inside the lock to keep the results consistent.

Commit:
raise IOError when dest does not exist

Commit:
assemble diff after unlock, check_mode does not lock file

Commit:
no locking when state=absent in check_mode

dagwieers commented:
What is the current status of this? An implementation would be widely useful to our users.

samdoran commented:
@dagwieers This is in my queue for review. I am working on other things at the moment, though.

samdoran commented:
@dagwieers Yup, planning to have it in for 2.8. I just don't know much about file locking, so I need to do some reading to be sure I understand the implementation well.

dagwieers commented:
IMO the easiest implementation that works for most (if not all) modules is to simply lock the file before you start reading it, and unlock it after you are done writing it. In the worst case (cornercases) you may exit before locking or unlocking, but both are fine even if not optimal, because if the module exits the file will be unlocked anyway. If there is already a lock active, the module waits until the lock is lifted until it times out.

A mechanism like this would work for every module that makes file modifications. Any optimization may introduce complexities introducing oversights introducing problems.

acalm commented:
@dagwieers @samdoran at the moment there's no default timeout set in neither lineinfile or common.file.FileLock, hence the default behavior is to wait forever until the lock . As discussed before: somewhere around 15~30s ought to be a sensible default for most implementations. However, I'm not entirely sure where this timeout should be set, should it be the calling module that provide a timeout value if needed or should common.file.FileLock include have: the most sensible default we can come up with?

samdoran commented:
@acalm I think the default timeout should be 15 seconds and set in the method itself. The caller can override it if needed.

samdoran commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,60 @@ class FileLock:
 
     @contextmanager
-    def lock_file(self, path, tmpdir, lock_timeout=None):
+    def lock_file(self, path, lock_timeout=None):
```
Set the default timeout here:
```suggestion
    def lock_file(self, path, lock_timeout=15):
```

samdoran commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,60 @@ class FileLock:
     Currently FileLock is implemented via fcntl.flock on a lock file, however this
     behaviour may change in the future. Avoid mixing lock types fcntl.flock,
-    fcntl.lockf and module_utils.common.file.FileLock as it will certainly cause
+    fcntl.flock and module_utils.common.file.FileLock as it will certainly cause
```
Does this need to be changed? It seems the documentation was originally warning against mixing `fcntl.flock` with `fcntl.lockf`, but now we have two references to the same method in the docs.

acalm replied:
Unintentional change, while trying lockf and then back to flock with search/replace. I'll fix it

samdoran commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,60 @@ class FileLock:
             self.unlock()
 
-    def set_lock(self, path, tmpdir, lock_timeout=None):
+    def set_lock(self, path, lock_timeout=None):
```
Set same default timeout here:
```suggestion
    def set_lock(self, path, lock_timeout=15):
```

samdoran commented on lib/ansible/module_utils/common/file.py:
```diff
@@ -49,57 +49,60 @@ class FileLock:
-        self.lockfd = open(lock_path, 'w')
+        self.lockfd = open(b_lock_path, 'ab')
+
+        if lock_timeout is None or lock_timeout < 0:
```
Since we are defaulting to `15` we can remove this `is None` check unless you think it's good to keep for sanity reasons.
```suggestion
        if lock_timeout < 0:
```

acalm replied:
I think it may be good for sanity, if someone passes None, since it might kind of suggest that timeout is disabled. I have no strong opinion though

samdoran reviewed (changes requested):
Just a few minor things. I like the overall approach. One thing I was unsure of is do we need to do any type of locking when actually writing the changes? I know we perform an atomic move as part of that operation so maybe not.

acalm commented:
The file will (and should) already be locked while the underlying functions like atomic_move perform changes, to prevent unexpected results. The underlying write functions does not have to be aware of the lock though, as long as other processes will check locks before trying to access/alter the file. It would of course be nice there was an implementation that made operations respect locks without the need for modules to use FileLock, but it would probably be a bit complex since the lock needs to be in place during both read and write of the file and the locked file descriptor would need to be passed around during the entire operation

Commit:
update changelog fragment

Commit:
set default timeouts, update docstring accordingly, correct FileLock comment/warning

acalm commented:
ready_for_review

dagwieers commented:
@acalm I tested **lineinfile** in one of my playbooks where I concurrently update a CSV file with facts from systems. I normally would do `serial: 1` on the play, and using your branch it is a lot faster. However I noticed that one of the entries our of 36 was not updated as expected.

This means that one of the threads was writing older information, which should not be possible.

dagwieers commented:
Can we maybe separate the file-lock changes from the **lineinfile** module ?
And add integration tests for the file-locking part ?

PS I would have expected the file-locking changes to **lineinfile** to be a lot more simple. Only lock on first read, and unlock on exit (which isn't even required).

dagwieers commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -263,15 +263,51 @@ def present(module, dest, regexp, line, insertafter, insertbefore, create,
+
+            msg, changed = check_file_attrs(module, changed, msg, attr_diff)
+
+    diff['before'] = to_native(b('').join(b_lines))
```
I would leave the `if module._diff:` in here.

acalm commented:
@dagwieers I'm not sure I have any useful integration tests, I've tested with an async task executed on multiple hosts (with ansible_connection=local) connecting to localhost writing to the same file, running ansible-playbook with --fork 12. I really need some pointers/help with getting these test right as I'm not sure if above do/test what I think it does and my experience writing integration tests for Ansible is rather limited

samdoran commented:
@acalm I agree with @dag that creating integration tests for this would be a good idea. You can create a new directory in `test/integration/targets/`, maybe `file_locking`. In there, you can create your playbook and a `runme.sh` script to execute your playbook. You can probably adapt what you already have for testing pretty easily.

A script would allow you to test different strategies and pass in any other settings like `--forks`.

Usually when you are writing a file you can verify the hash of the file matches what you expect using `stat` and `assert` modules.

Here are some useful examples of tests using scripts:
[test/integration/targets/hash](https://github.com/ansible/ansible/tree/ad386290b4d701712faafb91f17ad6fe7a361655/test/integration/targets/hash)
[test/integration/targets/include_import](https://github.com/ansible/ansible/tree/ad386290b4d701712faafb91f17ad6fe7a361655/test/integration/targets/include_import)

Please reach out on IRC in #ansible-devel with questions.

Commit:
simple integration tests for FileLock

dagwieers commented:
@acalm Well done, the integration test worked fine on my RHEL7 system !

dagwieers commented on test/integration/targets/file_lock/test_filelock.yml:
```diff
@@ -0,0 +1,36 @@
+---
+- hosts: lockhosts
+  gather_facts: false
+  become: true
```
This should not be needed.

dagwieers commented:
@acalm I can make it fail on my system if I e.g. increase the number of forks to 25 and create 100 instances. I ended up with 3 instances not having its entry added.

I am using the following changes:
```diff
diff --git a/test/integration/targets/file_lock/inventory b/test/integration/targets/file_lock/inventory
index 7c8c958..92eb6ae 100644
--- a/test/integration/targets/file_lock/inventory
+++ b/test/integration/targets/file_lock/inventory
@@ -1,2 +1,2 @@
 [lockhosts]
-lockhost[0:20] ansible_connection=local
+lockhost[0:100] ansible_connection=local
diff --git a/test/integration/targets/file_lock/runme.sh b/test/integration/targets/file_lock/runme.sh
index 4d50d75..3e9ccb2 100755
--- a/test/integration/targets/file_lock/runme.sh
+++ b/test/integration/targets/file_lock/runme.sh
@@ -2,5 +2,5 @@
 
 set -eux
 
-ansible-playbook test_filelock.yml -i inventory --forks 10 -v "$@"
+ansible-playbook test_filelock.yml -i inventory --forks 25 -v "$@"
 ansible-playbook test_filelock_timeout.yml -i inventory -v "$@"
diff --git a/test/integration/targets/file_lock/test_filelock.yml b/test/integration/targets/file_lock/test_filelock.yml
index e5595b9..280ce84 100644
--- a/test/integration/targets/file_lock/test_filelock.yml
+++ b/test/integration/targets/file_lock/test_filelock.yml
@@ -1,36 +1,25 @@
 ---
 - hosts: lockhosts
-  gather_facts: false
-  become: true
+  gather_facts: no
   vars:
     testfile: ~/ansible_testing/lock.test
   tasks:
-  - name: write inventory_hostname to testfile
+  - name: Write inventory_hostname to testfile concurrently
     lineinfile:
-      path: "{{ testfile }}"
-      line: "{{ inventory_hostname }}"
-      create: true
+      path: '{{ testfile }}'
+      line: '{{ inventory_hostname }}'
+      create: yes
       state: present
-    async: 60
-    poll: 0
-    register: write_waiter
 
-  - name: wait for write jobs to finish
-    async_status:
-      jid: "{{ write_waiter.ansible_job_id }}"
-    register: write_job_result
-    until: write_job_result.finished
-    retries: 60
-
-  - name: check testfile for inventory_hostname entries
+  - name: Check testfile for inventory_hostname entries
     lineinfile:
-      path: "{{ testfile }}"
-      line: "{{ inventory_hostname }}"
+      path: '{{ testfile }}'
+      line: '{{ inventory_hostname }}'
       state: present
     register: check_lockfile
 
-  - name: assert locking results
+  - name: Assert locking results
     assert:
       that:
-        - check_lockfile is not changed
-        - check_lockfile is not failed
+      - check_lockfile is not changed
+      - check_lockfile is not failed
```

dagwieers commented:
I also added this after the first **lineinfile** call:
```python
  - debug:
      msg: "File {{ testfile }} has {{ content.split('\n')|length }} lines for {{ ansible_play_batch|length }} instances"
    vars:
      content: "{{ lookup('file', testfile) }}"
    run_once: yes

  - name: Assert we get the expected number of lines
    assert:
      that:
      - content.split('\n')|length == ansible_play_batch|length
    vars:
      content: "{{ lookup('file', testfile) }}"
    run_once: yes
```

And I get something like:
```
TASK [debug] *************************************************************************************************************************************
ok: [lockhost0] => {
    "msg": "File ~/ansible_testing/lock.test has 81 lines for 101 instances"
}

TASK [Assert we get the expected number of lines] ************************************************************************************************
fatal: [lockhost0]: FAILED! => {
    "assertion": "content.split('\\n')|length == ansible_play_batch|length", 
    "changed": false, 
    "evaluated_to": false, 
    "msg": "Assertion failed"
}
```

dagwieers commented:
Surprisingly, if I run my tasks in async, it does work correctly.
```
TASK [debug] ***************************************************************************************
ok: [lockhost00] => {
    "msg": "File ~/ansible_testing/lock.test has 100 lines for 100 instances"
}

TASK [Assert we get the expected number of lines] **************************************************
ok: [lockhost00] => {
    "changed": false, 
    "msg": "All assertions passed"
}
```

dagwieers commented:
@calm Last night I reduced the code to the bare minimum (simple contextmanager, return file descriptor, no timeout) and it would still fail to work correctly. So I wondered if this is caused by: https://stackoverflow.com/questions/28470246/python-lockf-and-flock-behaviour

I then modified the **lineinfile** module to ensure it's only writing to the open file descriptor (and not opening the file again, or doing the atomic move). And that works perfectly fine. So the problem with the **lineinfile** module is that inside the with-block the same file is being opened again for reading and afterwards for writing, which breaks the existing lock even before it is finished.

Obviously we don't want to write directly to the file (for various reasons) and we need an alternative for the atomic move based on file descriptors. You can find my working example here: https://github.com/ansible/ansible/pull/52567

acalm commented:
@dagwieers I haven't tested this yet, but it might be possible to avoid this behaviour by not closing the file descriptor in FileLock and not doing UNLUCK, then just let it be closed/unlocked with the process dying, unless there is something weird with forking/async making them share file descriptors somehow, not sure why or even how. This might be vaguely related
man flock.2
`...
Locks  created  by flock() are associated with an open file description (see open(2)).  This means that dupli‐
       cate file descriptors (created by, for example, fork(2) or dup(2)) refer to the same lock, and this  lock  may
       be  modified  or released using any of these file descriptors.  Furthermore, the lock is released either by an
       explicit LOCK_UN operation on any of these duplicate file descriptors, or when all such file descriptors  have
       been closed.
...`

dagwieers commented:
If I do not replace the atomice_move, it fails. https://github.com/ansible/ansible/pull/52567/files#diff-f24c5c7f14dd803b9f885ec901ef9055R232

Even if I replace the destination file, before atomically replacing it, it still fails.

dagwieers commented:
So for completeness, the problem was not the `atomic_move()` but the fact that we were not updating the file descriptors for the processes waiting for a lock once the file was atomically moved.

acalm commented:
@dagwieers should we carry on with the changes here or should we do it in your PR?

Also, I was thinking, the class FileLock() should probably be renamed to just File or something like FileOperations since file related stuff in basic.py are to be be moved into common/file.py. The upside of renaming being that we could make more file operations (like unsafe_writes) "lock aware" I.E get the file descriptor behind the scenes without the module needing to pass the file descriptor, which, probably, would be nice. This is out of scope for this particular change, but the class renaming would make sense to do now as it will be more work changing it when it's used by more modules.

dagwieers commented:
I don't think we want to keep it as a class. Rationale is in my PR. BTW I spent 3h looking for another issue on Python3, for which I undid most of your **lineinfile** changes to ensure those were unrelated.

At this point, I don't know what's best.

Commit:
get common.file changes from dagwieers:file-locking #52567

acalm commented:
@dagwieers I checked out your changes to common.file (replaced open(file, 'ab+') with 'rb+' to avoid fd.seek(0)) in here and made some minor changes to lineinfile, it seems to work fine. However changing to fcntl.flock() on linux will still cause unexpected unlocks though

Commit:
FileLock -> lock

Commit:
pep8 fix

acalm commented:
@samdoran & @dagwieers I am confused, should I close this pr? It seems #52567 are pinned to the 2.8.0 milestone. I think the changes done to lineinfile here is rather nice, splitting data manipulation and file operations (see above review commends from @abadger ). I've pulled in the changes from #52567 here, if this should be closed perhaps we can pull the lineinfile changes from here into #52567 ?

Commit:
remove check_mode from common.file

Commit:
added missing extends_documentation_fragment

Commit:
use flock, do not change locking behaviour depending on platform

Commit:
pep8 fix

Commit:
avoid closing file descriptor anywhere in the process, let the os handle closing when process exits

Commit:
avoid messing with the file descriptor, locking mechanism will open in ab+ mode instead

Commit:
increase wait tick, do not close file descriptor

samdoran committed:
Fix PEP8 error

samdoran committed:
Ensure diff works properly when line is added

Create a copy of the list when passing it to a method to avoid manipulating the original list resulting in no diff being reported.

mattclay commented on test/integration/targets/file_lock/inventory:
```diff
@@ -0,0 +1,2 @@
+[lockhosts]
+lockhost[00:99] ansible_connection=local
```
```suggestion
lockhost[00:99] ansible_connection=local ansible_python_interpreter="{{ ansible_playbook_python }}"
```

samdoran committed:
Have the function change create a copy of the lines

Change the variable name used within the function so it's easire to understand that it's operating on a copy. This also makes it consistent with _absent_data_manipulator().

samdoran committed:
Specify ansible_python_interpreter to avoid weirdness when running in local mode

mattclay commented on test/integration/targets/file_lock/inventory:
```diff
@@ -1,2 +1,2 @@
 [lockhosts]
-lockhost[00:99] ansible_connection=local
+lockhost[00:99] ansible_connection=local ansible_python_interpreter="{{ ansible_playbook_python }}"
```
This is required by the test infrastructure to avoid using interpreter discovery in tests. Using interpreter discovery adds additional overhead and can result in selecting the wrong interpreter for tests.

samdoran replied:
Thanks for the explanation. I thought it was due to oddities around which Python is actually being used when running locally.

Commit:
close file descriptor when done

Commit:
rename task

Commit:
use lockf on LInux and flock on bsds

Commit:
pep8 fix

Commit:
locking only works for linux

Commit:
trailing whitespace

acalm commented:
closing this PR as work continues in @dagwieers PR #52567
